### DISCUSSION TITLE
Discussion: Inline pattern-matching on out arguments

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#5505] [Language Ideas] [answered]
Relax `out` argument syntax to accept a general pattern.
```diff
argument_value
-   : 'out' type identifier
+   : 'out' pattern
    | ...
    ;
```
If the pattern is not irrefutable, the calling method is required to return `bool`.
```cs
if (TryGet(args, out > 0))
```
->
```cs
if (TryGet(args, out var v) && v is > 0)
```
Note: `out var (x, y)` falls out of the pattern syntax.

**Comment by svick** [comment, marked as answer]
> If the pattern is not irrefutable, the calling method is required to return `bool`.

This part feels like too much magic to me, in that I think it would make it harder to understand what the expression actually does.

**Reply by alrz** [reply]
Thinking more about it, this is not backward compatible with `(out object v)`. That's not an irrefutable pattern (tests for null).

There will be some requirements on the method contract too. Maybe it's just better to rely on user-defined positional patterns for this sort of things.
